cartographer/internal/mapping_3d/local_trajectory_builder.cc

```diff
@@ -77,37 +77,39 @@ LocalTrajectoryBuilder::AddRangeData(const common::Time time,
     return nullptr;
   }
 
-  sensor::TimedPointCloud hits =
-      sensor::VoxelFilter(0.5f * options_.voxel_filter_size())
-          .Filter(range_data.returns);
+  if (num_accumulated_ == 0) {
+    first_pose_estimate_ = extrapolator_->ExtrapolatePose(time).cast<float>();
```

Comment:
What is the first_pose_estimate_ here? Shouldn't the time for the pose of the first point in the cloud be `time + common::FromSeconds(range_data.returns.front()[3])` ? 

I am confused: I thought that `common::FromSeconds(range_data.returns.front()[3])` is smaller than zero. So it seems to me that this line is causing the `ImuTracker` to be advanced to the time of the last point in the point cloud. When you later call `ExtrapolatePose()` with `t = time + common::FromSeconds(hit[3])` that should be before the time of the last point and hence cause a crash in the `ImuTracker` in my opinion. But since you are saying that you tested this by running it I must be misunderstanding something.

Reply:
Thanks for finding this bug, I didn't test the timed point cloud, I only use the same time for all points in same packet now. I will test the timed point cloud later.